type-hinting/README.md

```diff
@@ -0,0 +1,169 @@
+# Type Hinting
+
+Author: [@FrostyX](https://github.com/frostyx)
+
+This is not a case against type hinting (static types), this is a case
+against over-using them.
+
+## Good motivations
+
+- Type hints nicely complement docstrings. Docstrings are great for
+  describing the purpose of a function (or class, file) but they suck
+  for describing their arguments and return value types because they
+  can be untrue. Type hints nicely complement this.
+- Type hints can reveal some errors
+
+
+## Bad motivations
+
+### A bulletproof solution
+
+Static types are not a bulletproof solution for revealing type-related
+bugs. IIRC some study found that ~80% of type-related errors are
+caused by `None` and Mypy won't save us there - it can find many
+issues, but I can still introduce tracebacks or unexpected bugs that
+Mypy doesn't warn about:
+
+Produces traceback:
+
+```python
+def hello(user: Optional[User]) -> str:
+    return "Hello " + getattr(user, "name")
+
+print(hello(None))
+```
+
+Returns `Hello None`:
+
+```python
+def hello(user: Optional[User]) -> str:
+    return "Hello {0}".format(user)
+
+print(hello(None))
+```
+
+Produces traceback:
+
+```python
+def hello(user: dict) -> str:
+    return "Hello {0}".format(user["name"])
+
+print(hello({}))
+```
+
+### Performance
+
+Type hinting in python
+[doesn't bring any performance improvements][mypy-performance],
+it is used merely for linting.
+
+
+## Disadvantages
+
+### One does not simply know the type
+
+_Insert the Boromir meme_
+
+Duck typing is IMHO one of the best Python features ...
+Normally, we are not required to know and say exactly what type is
+expected but rather how it is supposed to behave.
+
+I may intend to write a function that filters a list somehow:
+
+```python
+def odd(numbers: list):
+    return [x for x in numbers if x%2 == 0]
+```
+
+And expect the following usage
+
+```python
+print(odd([1, 2, 3, 4]))
+```
+
+Later, somebody want to use
+
+```python
+print(odd(range(1, 5)))
+print(odd(reversed([1, 2, 3, 4])))
+```
+
+and Mypy complains.
```

Comment:
Duck typing is a great advantage of python indeed and python static typing (hopefully) is trying not to kill it.

You are telling here that you are expecting `list` type which is explicit type. If you want to benefit from duck typing, you need to use a type which tells mypy how it's supposed to behave. In this case, you need to tell that the `number` is some kind of parameter that can be iterated. 
For this case, `typing` offers an `Iterable` type. So this following code will pass even with `--strict` option:

```py
def odd(numbers: Iterable[int]) -> list[int]:
    return [x for x in numbers if x % 2 == 0]
```

Of course this is a trivial example and in reality you often need more complex types and structures that tells mypy how it is supposed to behave. But this is already quite an advanced topic. In the vast majority of more complex cases an [Union](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#union-types) will suffice.  Of course, sometimes even that may not be enough. In these cases, I think it's perfectly fine to type the part with [Any](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-any-type), which says that the variable/function... can be anything. Sure you can create complex types which will be correct but this is overkill... it makes sense for startups with experienced mypy guys... or for masochist :D